### PR TITLE
Adding Feature: Snapshot

### PR DESCRIPTION
**Pull request by ydlamba**
**To Do -**
- [x] Add button in export modal
- [x] Add copy to clipboard option
- [x] Show generated URL in modal
...

### CODE REVIEW

File: src/components/app-shell.tsx [hunk 1]
```diff
@@ -21,6 +21,7 @@ export default class AppShell extends React.Component<Props> {
           <Route path="/gist/:mode/:username/:id/:revision/:filename" component={App} />
           <Route path="/examples/:mode/:example_name" component={App} />
           <Route path="/custom/:mode" component={App} />
+          <Route path="/shared/:spec" component={App} />
```
Comment (domoritz) [inline src/components/app-shell.tsx]: How do we know whether it's Vega or Vega-Lite?

File: src/components/header/renderer.tsx [hunk 2]
```diff
@@ -248,6 +249,12 @@ class Header extends React.Component<Props & { history: any }, State> {
         {'Export'}
       </div>
     );
+    const saveButton = (
+      <div className="header-button" onClick={() => this.props.saveSpec(true)}>
+        <Save className="header-icon" />
+        {'Snapshot'}
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: Wasn't the plan to put this in the sharing modal?

File: src/components/app-shell.tsx (hunk 1 above)
Comment (domoritz) [reply src/components/app-shell.tsx]: rename to just `share`

**Comment by domoritz** [comment]
- [ ] Add compression

**Comment by ydlamba** [comment]
Pull request has been rebased and Now, I'm working on the snapshot feature.

**Comment by domoritz** [comment]
- When the sharing modal is open, always generate the URL. Don't wait. 
- Instead of a text input, let's use a URL so the user can see the changes when they check fullscreen. Preview is then just a matter of clicking on the button
- No need to the big button. 
- make the copy to clipboard more visible. Add some text for it! Never just rely on icons.

**Comment by domoritz** [comment]
- add a byte counter (and show a warning when we are over 2000)
- make the link look like a link. use underline and remvoe the grey box
- but the copy button in a prominent location
- write "copy to clipboard". copy by itself is not enough

File: src/components/header/renderer.tsx [hunk 3]
```diff
@@ -479,6 +525,64 @@ class Header extends React.Component<Props, State> {
       </div>
     );
 
+    const copiedClass = 'copied' + (this.state.copied ? ' visible' : '');
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: Use string template

File: src/components/header/renderer.tsx [hunk 4]
```diff
@@ -479,6 +525,64 @@ class Header extends React.Component<Props, State> {
...
+            </div>
+          </Clipboard>
+          <div className={copiedClass}>Copied!</div>
+        </div>
+        <div className="byte-counter">
+          Bytes: {this.state.generatedURL.length}{' '}
+          <span className="warning">
+            {this.state.generatedURL.length > 2083 ? (
+              <span>
+                Warning:{' '}
```
Comment (domoritz) [inline src/components/header/renderer.tsx]: add space between : and content

File: src/components/input-panel/spec-editor/renderer.tsx [hunk 5]
```diff
@@ -90,7 +91,8 @@ function debounce(func, wait, immediate?) {
 interface Props {
   autoParse?: boolean;
   format?: boolean;
-  history;
+  history: any;
+  match: any;
```
Comment (domoritz) [inline src/components/input-panel/spec-editor/renderer.tsx]: what's this for?

File: src/components/header/index.css [hunk 6]
```diff
@@ -433,3 +527,18 @@
     margin-right: -105px;
   }
 }
+
+/* Scrollbar Styles */
```
Comment (domoritz) [inline src/components/header/index.css]: These are global styles, right? Move them somewhere else then.

File: src/components/header/index.css [hunk 7]
```diff
@@ -78,6 +78,7 @@
   position: relative;
   background: white;
   width: 80%;
+  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
```
Comment (domoritz) [inline src/components/header/index.css]: Why is this necessary?

File: src/components/app.css [hunk 8]
```diff
@@ -77,6 +77,13 @@ html {
   overflow: visible !important;
 }
 
+button,
+input,
+textarea,
+select {
+  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
```
Comment (domoritz) [inline src/components/app.css]: Again? I thought we set this on the body.

File: src/components/header/renderer.tsx (hunk 4 above)
Comment (ydlamba) [reply src/components/header/renderer.tsx]: There is a space in the final rendering.

**Comment by ydlamba** [comment]
PR has been updated with final changes.

File: src/components/header/index.css (hunk 6 above)
Comment (domoritz) [reply src/components/header/index.css]: They are still here

File: src/components/header/index.css [hunk 9]
```diff
@@ -413,6 +403,110 @@
...
+  vertical-align: middle;
+  line-height: 25px;
+  white-space: nowrap;
+}
+
+.exported-url a::selection {
+  background: rgba(0, 0, 0, 0.15);
+}
+
+.share-buttons,
```
Comment (domoritz) [inline src/components/header/index.css]: We are still not sharing the style between the data viewer and the modals, right?
